src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -24,53 +26,83 @@
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic;
 
-internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
+[Export(typeof(IRazorSemanticTokensInfoService)), Shared]
+[method: ImportingConstructor]
+internal class RazorSemanticTokensInfoService(
+    IRazorDocumentMappingService documentMappingService,
+    LanguageServerFeatureOptions languageServerFeatureOptions,
+    IRazorLoggerFactory loggerFactory,
+    ITelemetryReporter? telemetryReporter)
+    : IRazorSemanticTokensInfoService
 {
     private const int TokenSize = 5;
 
-    private readonly IRazorDocumentMappingService _documentMappingService;
-    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
-    private readonly RazorLSPOptionsMonitor _razorLSPOptionsMonitor;
-    private readonly IClientConnection _clientConnection;
-    private readonly ILogger _logger;
+    private readonly IRazorDocumentMappingService _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
```

Comment:
>_documentMappingService [](http://example.com/codeflow?start=50&length=23)

Nit: what do you think of suggestions here 

[https://stackoverflow.com/questions/77556444/null-checking-with-primary-constructor-in-c-sharp-12](https://stackoverflow.com/questions/77556444/null-checking-with-primary-constructor-in-c-sharp-12)

Couple of interesting thoughts there. I personally like the idea of re-using primary constructor parameter name for class field name to shadow it. That way you won't have both "documentMappingService" and "_documentMappingService" in your completion list. Also that helper methods seems nice, though it becomes another "magic" method someone not familiar with codebase won't immediately understand when looking at the code. So not sure about the helper, but re-using parameter names seems nice.

Reply:
Probably something we should discuss, and maybe vote on, in a meeting early next year. Personally I'm a fan of having separate fields because a) they're readonly, and b) its obvious in a PR diff whether something is a field because of the underscore. If a primary constructor parameter is captured in a field, then the compiler will warn if the parameter is used in a method body, so its not really possible to mix the two styles.

As for the extension method, I don't have any issue with it. We already have one, `.AssumeNotNull()` which we could use, though it throws a different exception type. Don't know if that matters. My real personal preference there is to get rid of the null checks altogether and just rely on the MEF composition exceptions.

Reply:
I don't have strong feelings here and I also like having underscores for readability, but then should we *always* have private (readonly when appropriate) field starting with underscore for primary constructor parameters used in class methods? Definitely something we should agree on to be consistent. Seems like half of the clean-up from primary constructors came from getting rid of fields, then of course naming becomes a bit confusing, especially in long files when you are looking at a name and you don't know if it's a local defined somewhere above in the method or primary constructor parameter. If we do have fields for each parameter, do we even need to switch to primary constructors?

I'm good with relying on MEF composition exceptions in MEF parts, the code wouldn't even get to throwing ANE exception in that case, so the check is pointless. 

Might be nice to have ANE thrown instead of IOE, so maybe adding ArgumentNotNull would be nice in places where we keep null checking for arguments. Not asking to add in this PR :). I was looking at this mostly in the context of switching to primary constructors and having each field initializer doing the whole "name ?? throw new ANE(nameof(name))" everywhere to see if there is something more succinct we could do.